### PR TITLE
Convert TOC to use cache map

### PR DESCRIPTION
**Pull request by dbutenhof**
PBENCH-1192

This has been on my wishlist for a while, but was blocked by not actually having a usable cache. PR #3550 introduces a functioning (if minimal) cache manager, and this PR layers on top of that. The immediate motivation stems from an email exchange regarding Crucible, and the fact that Andrew would like (not surprisingly) to be able to access the contents of an archived tarball. Having TOC code relying on the Pbench-specific `run-toc` Elasticsearch index is not sustainable.

Note that, despite #3550 introducing a live cache, this PR represents the first actual use of the in-memory cache map, and some adjustments were necessary to make it work outside of the unit test environment.

Only the final commits here represent the TOC changes: most are from the prior PR.

### CODE REVIEW

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_cache_manager.py:951]: Why did you remove this test?  I see that you've added a similar test in `test_datasets_contents.py`, but that's for a different "unit".

File: lib/pbench/test/unit/server/test_datasets_contents.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_contents.py:66]: Whatever you do about replacing `Path("")` with either `Path(".")` or `Path()`, you should do the same here, and at line 149, etc.

File: lib/pbench/test/unit/server/test_datasets_contents.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_contents.py:64]: I don't see anywhere where you use the value of `key` when it is not `"subdir1"`...is there somewhere, or are the False-y cases redundant?

File: lib/pbench/test/unit/server/test_datasets_contents.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_contents.py:98]: Given that the call to `find_entry()` is going to raise an exception, do these mocks actually get called?  (Ditto lines 80 & 81.)

**Review by webbnh — COMMENTED** [review]
I have a bunch of comments, most of which are small.  However, I'm concerned that you've removed the test case for a symlink which references an existing directory outside the dataset result.  Also, it's not obvious whether a `CacheType.OTHER` should have a `"size"` key in the "details", and the code is not consistent on that point (which means we're probably missing a test case).  And, I think some of the `CacheMapEntry`/`CacheMap` type hints are now wrong.

File: lib/pbench/server/api/resources/datasets_contents.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_contents.py:70]: I was really bothered by `location` values of `"."` in the map for the root directory, and went to (perhaps too much) effort to stamp them out.

But, yeah; `Path("")` isn't just "equivalent to" `Path(".")` ... the `str()` of `Path("")` *is* `"."`. I got sucked into all this as a consequence of "invisibilizing" the self-named root directory and building the cache map from there, while really trying to avoid "." on the output paths.

File: lib/pbench/server/api/resources/datasets_contents.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_contents.py:78]: Maybe. We know where it was generated, and `APIInternalError` has a stack trace. Building strings around `str(e)` often results in unwieldy and redundant messages which I consciously avoided here. Most system exception messages are also pretty clear as-is. Maybe not `KeyError` though, and this *is* a `except Exception`. 😆

File: lib/pbench/server/api/resources/datasets_contents.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_contents.py:155]: Yeah, I remember throwing the `CacheType.OTHER` into the tuple above and deciding not to worry about it rather than make a special case. But I don't think "size" is a meaningful concept here aside from `FILE`, so maybe I need to break up the `dict` up above.

File: lib/pbench/server/api/resources/datasets_contents.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_contents.py:160]: Want? Need? Care? This was way below the cut line when I was struggling to refactor the code to get it working, and later I didn't have the energy to go back and think about it again. I'm still not convinced it's worth worrying about, nor do I like the long and often redundant messages I've seen us build elsewhere. But ... I don't know.

File: lib/pbench/server/cache_manager.py
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py:168]: What I found "surprising" was `https://<node>/api/v1/datasets/<id>/contents/.`; and it bugged me and I took to executing dots on sight. I suspected at the time that I went too far and that you'd probably call me on it, but I was worn out and did it anyway!

File: lib/pbench/server/cache_manager.py
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py:611]: You're probably right. I found the naming and use of those two highly confusing and somewhat poorly defined. I considered getting rid of the `dict` looseness here and building real `CacheMap` and `CacheMapEntry` dataclasses, but didn't quite take it that far. I think that would be easier to follow since at least the IDE cares (a little).

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py:975]: We want _either_ for the `"children"` key to exist _or_ for the file type to be something other than `DIRECTORY` but not **both**...so `xor`.  😁  (Sorry that the expression wraps. 😢  I didn't think you'd actually be interested so I didn't check....  😆)

**Review by dbutenhof — COMMENTED** [review]
OK; I found a bunch of duplicated lines from a weird rebase that I need to fix...

**Comment by webbnh** [comment]
If you still have the original commits available (like if the hash IDs happen to be in your terminal scrollback), you should consider starting over from your un-re-based branch.

You might have better luck if you squash your branch (particularly the commits which come from the other PR) before rebasing.

**Comment by dbutenhof** [comment]
> If you still have the original commits available (like if the hash IDs happen to be in your terminal scrollback), you should consider starting over from your un-re-based branch.
> 
> You might have better luck if you squash your branch (particularly the commits which come from the other PR) before rebasing.

Yeah, in cases like this I would have been better off squashing both before the rebase, I suspect. But aside from git deciding to duplicate some lines without showing them as conflicts, it wasn't terrible and I definitely wouldn't want to start over! Just a few more segments to clean up: and scanning through the diffs on GitHub pointed out a few comments I'd overlooked after the rebase scramble. I should be able to clean this up "quickly"(ish) tomorrow morning...

**Comment by webbnh** [comment]
And, for future reference, when you rebase a branch of a branch onto main when the base-branch has been merged, you also have the option of specifying the branching-off place (so, you rebase just the last commit or three, instead of trying to rebase the whole branch hoping that Git will drop most of the commits because they are already in the target).

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_cache_manager.py:660]: The `strict` parameter is never used, so it should retain its leading underscore.

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_cache_manager.py:521]: Since `subdir15` is not expected to exist (it's not...right?), maybe we should give it a more expressive name?  (Like "nonexistent_dir"?)

File: lib/pbench/server/indexing_tarballs.py
Comment (webbnh) [inline lib/pbench/server/indexing_tarballs.py:379]: I assume that you've already noted this, but I think you removed this change and replaced it with a context manager which produced a nice result...hopefully you can restore that code.

File: lib/pbench/server/api/resources/datasets_contents.py
Comment (webbnh) [inline lib/pbench/server/api/resources/datasets_contents.py:101]: Should this list now include `CacheType.BROKEN`?

File: lib/pbench/server/api/resources/datasets_contents.py
Comment (webbnh) [inline lib/pbench/server/api/resources/datasets_contents.py:155]: Should there be a `BROKEN` branch here, too?

**Review by webbnh — COMMENTED** [review]
I realize that this PR is currently a draft, but here are some things I found.

File: lib/pbench/server/cache_manager.py
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py:102]: I suppose you're right: I always liked how Javadoc allowed adding individual descriptions for Enum values by doing the equivalent: but Python doesn't association docstrings with the values (e.g., on IDE flyover), so it's at best wasted space. 😦

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_cache_manager.py:521]: I struggled with the naming conventions here. Riya chose to give hierarchical names, which are useful in one sense, but I had to keep coming back here to figure out which were good/bad/etc. I did consider changing the names to record the function rather than just the hierarchy, and I just didn't want to get into making sure I'd tracked everything down.

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_cache_manager.py:523]: I know: I did that in a few cases, but didn't try to do it consistently. Mostly the repeated variable interspersed with the creations made it harder (for me, at least) to read ... and I ended up doing that a lot. You're right that consistently dropping the path concatenation where it's not necessary might help readability further, so I'll take a look at that.

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_cache_manager.py:658]: No, it shouldn't, because Python doesn't like a call with `strict=True` when there's no `strict` argument defined.

File: lib/pbench/server/api/resources/datasets_contents.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_contents.py:159]: Now my reviewer is asking me to add broken code. Who woulda guessed? 😆

File: lib/pbench/server/indexing_tarballs.py
Comment (dbutenhof) [reply lib/pbench/server/indexing_tarballs.py]: Yeah; again, nothing was dropped, but the rebase restored some explicit `lock` references that are no longer relevant. The code ran just fine, though, because it also restored the leading `lock = None`. It was wrong ... but at least it was consistent.

File: lib/pbench/test/unit/server/test_datasets_contents.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_contents.py:70]: Can you just confirm that the expression at line 70 is correct?

If `key` is `""` then we use `""` instead of `file.name`, but, if `key` is `"."`, we don't use `""` -- since I thought that `""` and `"."` were supposed to be equivalent, I'm not understanding this difference.  (E.g., why is this expression different from the one at line 81?)

File: lib/pbench/test/unit/server/test_datasets_contents.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_contents.py:133]: Isn't line 133 redundant?  (Doesn't `query_get_as()` do that check using the third parameter?)

Ditto line 225.

**Review by webbnh — COMMENTED** [review]
Looks generally good.  There are a couple of small things to be attended to (there are some lingering repeated lines and similar, et al.).

File: lib/pbench/server/cache_manager.py
Comment (webbnh) [reply lib/pbench/server/cache_manager.py:659]: Duh:  `path` is a `Path`, and it's not `Optional`, so...nevermind!  🥴

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_cache_manager.py:536]: Is that really how we want to express this?  Wouldn't `Path("./f1411.txt")` or `Path("f1411.txt")` be better?

File: lib/pbench/test/unit/server/test_datasets_contents.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_contents.py:75]: If you make a change for some other reason, consider moving the assignment of `name` from line 86 to where this line can reference it instead of replicating it.

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py]: Yeah, let's not track them all down, but we now know about this one....  😁

File: lib/pbench/test/unit/server/test_datasets_contents.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_datasets_contents.py]: Thanks for the docstrings -- they help a lot (and that's good enough).

However...

> because the `ARCHIVE` tree hasn't been filled in or mocked

If we're counting on the unit test looking at the real file system and finding it empty, that's a little weak.  It would (have) be(en) better if there were an explicit mock which drove the behavior that we want to see.

**Review by webbnh — APPROVED** [review]
Looks good.  Just a few nits to consider.

File: lib/pbench/test/unit/server/conftest.py
Comment (webbnh) [inline lib/pbench/test/unit/server/conftest.py:1]: The changes to this file appear to be reverting two of the changes that I made in #3557...is that intentional?

**Review by webbnh — CHANGES_REQUESTED** [review]
I just notice this:  please confirm before merging.

**Review by webbnh — APPROVED** [review]
Good to go (unless you want to do something for the two remaining open conversations).

**Review by dbutenhof — COMMENTED** [review]
Dang; I had responses pending, and I thought I'd submitted them. Sigh.

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py]: `Path(".") / "f1412_sym"`, `Path("./f1412_sym")`, and `Path("f1412_sym")` all create the same `PosixPath` object; and, therefore, when a symlink is made from it, the contents of the link file are set to `f1412_sym` without a leading dot-slash.

So, if we want a path starting with a dot in the reference, we have to use a _string_ for the first argument to `os.symlink()` and not that `Path` value.

**Comment by webbnh** [comment]
> 3 of 4 checks passed

It looks like maybe you merged this before the tests were done, and GitHub permanently captured that information...interesting!  🤔

File: lib/pbench/test/unit/server/conftest.py
Comment (webbnh) [reply lib/pbench/test/unit/server/conftest.py:1]: Those lines were changed on the `main` branch; the other lines were changed identically (to the best of my ability 😉) in both branches.  So, I think you ended up with your version of the file instead of the version from `main` or a "merged" version.

**Comment by dbutenhof** [comment]
> > 3 of 4 checks passed
> 
> It looks like maybe you merged this before the tests were done, and GitHub permanently captured that information...interesting! 🤔

Yikes -- I hadn't actually meant to do that, and I think I just assumed it was done. 😦 

But I don't think it's "captured" -- even though you have to hit another button to open the test details pane, it's still there and the summary will presumably update when the tests are done.

**Pull request merged**
